matrix-sdk-android/src/main/java/im/vector/matrix/android/internal/database/AsyncTransaction.kt

```diff
@@ -19,27 +19,36 @@ import io.realm.Realm
 import io.realm.RealmConfiguration
 import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.isActive
+import kotlinx.coroutines.sync.Semaphore
+import kotlinx.coroutines.sync.withPermit
 import kotlinx.coroutines.withContext
 import timber.log.Timber
 
-suspend fun <T> awaitTransaction(config: RealmConfiguration, transaction: suspend (realm: Realm) -> T) = withContext(Dispatchers.Default) {
-    Realm.getInstance(config).use { bgRealm ->
-        bgRealm.beginTransaction()
-        val result: T
-        try {
-            val start = System.currentTimeMillis()
-            result = transaction(bgRealm)
-            if (isActive) {
-                bgRealm.commitTransaction()
-                val end = System.currentTimeMillis()
-                val time = end - start
-                Timber.v("Execute transaction in $time millis")
-            }
-        } finally {
-            if (bgRealm.isInTransaction) {
-                bgRealm.cancelTransaction()
+// TODO: Make this more of a global realm (or database) thing.
```

Comment:
I do not understand. @ganfra could you check the change to this file please? Thanks

Reply:
With this change, all calls to `awaitTransaction` will be synchronised, but if transactions are started without using `awaitTransaction` will run concurrently. So ideally everywhere should either always use `awaitTransaction` or use the `Semaphore`.

Although, if `awaitTransaction` is already consistently used everywhere for transactions, then that comment should be removed.